src/session-pool.js

```diff
@@ -109,221 +113,178 @@ function SessionPool(database, options) {
+      autostart: false,
+      maxWaitingClients: this.options.maxWait,
+      evictionRunIntervalMillis: this.maxIdleResourceTimeout,
+      numTestsPerEvictionRun: this.options.minReads,
+      idleTimeoutMillis: this.maxIdleResourceTimeout,
+      testOnBorrow: true,
+    }
+  );
 
-  return Promise.all(requests);
-};
+  this.writePool = Pool.createPool(
+    {
+      create: this.createWriteSession.bind(this),
+      destroy: this.destroySession.bind(this),
+      validate: validateSession.bind(this),
+    },
+    {
+      max: this.options.maxWrites,
+      min: this.options.minWrites,
+      acquireTimeoutMillis:
+        this.options.acquireTimeout === Infinity
+          ? 0
+          : this.options.acquireTimeout,
+      autostart: false,
+      evictionRunIntervalMillis: this.maxIdleResourceTimeout,
+      maxWaitingClients: this.options.maxWait,
+      numTestsPerEvictionRun: this.options.minWrites,
+      idleTimeoutMillis: this.maxIdleResourceTimeout,
+      testOnBorrow: true,
+    }
+  );
 
-/**
- * Returns stack traces for sessions that have not been released.
- *
- * @return {string[]}
- */
-SessionPool.prototype.getLeaks = function() {
-  return Array.from(this.traces_.values()).map(SessionPool.formatTrace_);
-};
+  this.readPool.on('factoryCreateError', this.emit.bind(this, 'error'));
+  this.readPool.on('factoryDestroyError', this.emit.bind(this, 'error'));
+  this.writePool.on('factoryCreateError', this.emit.bind(this, 'error'));
+  this.writePool.on('factoryDestroyError', this.emit.bind(this, 'error'));
 
-/**
- * Retrieve a read session.
- *
- * @returns {Promise<Session>}
- */
-SessionPool.prototype.getSession = function() {
-  return this.acquireSession_(READONLY);
-};
+  EventEmitter.call(this);
+}
 
-/**
- * Retrieve the write session.
- *
- * @returns {Promise<Session>}
- */
-SessionPool.prototype.getWriteSession = function() {
-  return this.acquireSession_(READWRITE);
-};
+util.inherits(SessionPool, EventEmitter);
 
 /**
- * Checks to see if the pool is full by calculating the current size and the
- * pending number of sessions being created.
+ * This method validates the session. The session is invalid if the session is idle for more than the maxIdleResourceTimeout
+ * This is internally used by generic-pool and hence there is no need to make it a method of the class.
  *
- * @return {boolean}
+ * @param {Object} session
  */
-SessionPool.prototype.isFull = function() {
-  return this.size() + this.pendingCreates_ >= this.options.max;
-};
-
+function validateSession(session) {
+  const now = Date.now();
+  if (!session || !session.lastUsed) {
+    return Promise.resolve(false);
+  }
+  const flag = now - session.lastUsed < this.maxIdleResourceTimeout;
+  session.lastUsed = now;
+  return Promise.resolve(flag);
+}
 /**
  * Opens the pool, filling it to the configured number of read and write
  * sessions.
  *
- * @emits SessionPool#open
  * @return {Promise}
  */
 SessionPool.prototype.open = function() {
```

Comment:
This function looks to be largely untested. I only see a test for the `isOpen` flag. Can we add tests for starting the read/write pools and a test to ensure the correct values are passed to the `setTimeout` call

Reply:
update the test case